### PR TITLE
PLANNER-1172: Fix add tenant issues

### PR DESCRIPTION
**Pull request by Christopher-Chianelli**

### CODE REVIEW

File: employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/common/StringListToSkillSetConverter.java [hunk 1]
```diff
@@ -103,17 +108,14 @@ private void updateSkillMappings(List<Skill> skillList) {
...
     }
 
     private Promise<List<Skill>> getSkillList() {
         return new Promise<>((resolve, reject) -> SkillRestServiceBuilder.getSkillList(tenantStore.getCurrentTenantId(), FailureShownRestCallback
-                .onSuccess(newSkillList -> {
-                    resolve.onInvoke(newSkillList);
-                })));
+                                                                                                                                                 .onSuccess(newSkillList -> {
+                                                                                                                                                     resolve.onInvoke(newSkillList);
+                                                                                                                                                 })));
```
Comment (ge0ffrey) [inline employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/common/StringListToSkillSetConverter.java]: @yurloc @rsynek Are any of you guys a fan of the way method parameters get idented by the general kie style?

Comment (Christopher-Chianelli) [reply employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/common/StringListToSkillSetConverter.java]: I am not, to be honest

File: employee-rostering-gwtui/src/main/resources/org/optaweb/employeerostering/gwtui/client/admin/AdminPage.html [hunk 2]
```diff
@@ -1,5 +1,32 @@
...
+                <button data-field="add-tenant-button">Add Tenant</button>
+              </div>
+            </div>
+          </div>
+        </div><!-- /col -->
+      </div>
+    <table data-field="table" class="table table-hover kie-tablelist">
+      <thead>
+        <tr>
+          <!-- TODO: i18n -->
```
Comment (ge0ffrey) [inline employee-rostering-gwtui/src/main/resources/org/optaweb/employeerostering/gwtui/client/admin/AdminPage.html]: I suspect that these TODO: i18n comments should be just about everywhere in UI, which will just clutter the code.
Let's not keep copy pasting them to some area - and leave them out.

I'l also make a i18n issue in jira, so we can make work of putting all UI strings in i18n property files.

Comment (ge0ffrey) [reply employee-rostering-gwtui/src/main/resources/org/optaweb/employeerostering/gwtui/client/admin/AdminPage.html]: Wait, a sec, we have i18n support thanks to your work: https://github.com/kiegroup/optaweb-employee-rostering/pull/13/files

Ok, in that case, fix this to use that i18n todo :)

File: employee-rostering-shared/src/main/java/org/optaweb/employeerostering/shared/roster/RosterState.java [hunk 3]
```diff
@@ -177,14 +179,11 @@ public void setDraftLength(Integer draftLength) {
         this.draftLength = draftLength;
     }
 
+    @JsonIgnore
     public Integer getPublishLength() {
         return publishLength;
     }
 
-    public void setPublishLength(Integer publishLength) {
```
Comment (ge0ffrey) [inline employee-rostering-shared/src/main/java/org/optaweb/employeerostering/shared/roster/RosterState.java]: Is the RosterState ever send to the server? I wasn't sure if jackson doesn't need the setter to unmarshal it?

File: employee-rostering-server/src/main/java/org/optaweb/employeerostering/server/tenant/TenantRestServiceImpl.java [hunk 4]
```diff
@@ -60,6 +68,36 @@ public Tenant addTenant(RosterState intialRosterState) {
...
+        deleteAllEntitiesBelongingToTenantOfType(tenantId, RosterState.class, "RosterState.find");
+        entityManager.remove(tenant);
+        return true;
+    }
+
+    private <T extends AbstractPersistable> void deleteAllEntitiesBelongingToTenantOfType(Integer tenantId, Class<T> type, String queryName) {
+        List<T> entitiesBelongingToTenant = entityManager.createNamedQuery(queryName, type).setParameter("tenantId", tenantId)
+                                                         .getResultList();
+        for (T entity : entitiesBelongingToTenant) {
+            entityManager.remove(entity);
```
Comment (ge0ffrey) [inline employee-rostering-server/src/main/java/org/optaweb/employeerostering/server/tenant/TenantRestServiceImpl.java]: This is slow.

Add a JPA-QL per entity type that does "delete where tenantId = $tenantId"

File: employee-rostering-shared/src/main/java/org/optaweb/employeerostering/shared/roster/RosterState.java (hunk 3 above)
Comment (Christopher-Chianelli) [reply employee-rostering-shared/src/main/java/org/optaweb/employeerostering/shared/roster/RosterState.java]: When you create a Tenant, you need to supply the initial `RosterState`. I added an `@JsonIgnore` on the getter since publish length is not sent (since it is non-modifiable)

File: employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/common/StringListToSkillSetConverter.java (hunk 1 above)
Comment (rsynek) [reply employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/common/StringListToSkillSetConverter.java]: @ge0ffrey you could not have picked a better illustration to get the answer "No!" :) I even wonder why we have such a thing as KIE code style. Why not simply taking Oracle Java conventions? In contrast to KIE code style, it's already available for every existing IDE and maintenance cost for the team is none.

Comment (yurloc) [reply employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/common/StringListToSkillSetConverter.java]: If this is formatted with KIE Java Conventions then the Eclipse version of the formatting rules is broken. IDEA doesn't do this. Please do not commit this.

Also, this is not related to parameter indentation (in this case, all parameters are on the same line).

File: employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/spot/SpotListPanel.java [hunk 5]
```diff
@@ -34,20 +35,20 @@
...
 import org.optaweb.employeerostering.gwtui.client.tenant.TenantStore;
-import org.optaweb.employeerostering.gwtui.client.util.CommonUtils;
 import org.optaweb.employeerostering.gwtui.client.util.PromiseUtils;
 import org.optaweb.employeerostering.shared.spot.Spot;
 import org.optaweb.employeerostering.shared.spot.SpotRestServiceBuilder;
 
 @Templated
 public class SpotListPanel implements IsElement,
-        Page {
+                           Page {
```
Comment (yurloc) [inline employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/spot/SpotListPanel.java]: `Page` should be aligned with `IsElement`. @Christopher-Chianelli I don't want to block the PR with this one but sooner we get Eclipse and IDEA formatters in sync the sooner we'll get clean PRs without unnecessary whitespace changes,

File: employee-rostering-server/src/main/java/org/optaweb/employeerostering/server/tenant/TenantRestServiceImpl.java (hunk 4 above)
Comment (yurloc) [reply employee-rostering-server/src/main/java/org/optaweb/employeerostering/server/tenant/TenantRestServiceImpl.java]: Maybe consider using Criteria API? https://stackoverflow.com/questions/23846618/bulk-delete-in-jpa-2-1-using-the-criteria-api

File: employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/spot/SpotListPanel.java (hunk 5 above)
Comment (yurloc) [reply employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/spot/SpotListPanel.java]: @Christopher-Chianelli Thanks for confirmation. Do you think you can fix the Eclipse configuration and submit a PR to `droolsjbpm-build-bootstrap?

Comment (Christopher-Chianelli) [reply employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/spot/SpotListPanel.java]: @yurloc It appears IDEA can use Eclipse formatter, but not the other way around: https://intellij-support.jetbrains.com/hc/en-us/community/posts/206575559-How-to-Export-Code-Formatting-Configurations-to-Eclipse. I cannot guarantee the formatters will be one-to-one, but I try to make it as one-to-one as possible. We should consider adding a format checker in addition to the already existing checkstyle checker (https://github.com/kiegroup/droolsjbpm-build-bootstrap/blob/f8000107b3d94c2a05e46beefa0b9cbedd2850c1/pom.xml#L590). See https://github.com/revelc/formatter-maven-plugin (this uses the eclipse formatter, did not find one that uses the IDEA formatter yet).

Comment (Christopher-Chianelli) [reply employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/spot/SpotListPanel.java]: @yurloc Can confirm it is impossible to make `Page` be aligned with `IsElement` in the Eclipse formatter. "Line Wrapping > 'implements' cause > Indentation policy" controls the indent. There are three options: "Default indentation", which causes it to be indented at column 8 (2 standard indents), "Indent on column", which causes it to be indented on "implements" and "Indent by 1", which causes it to be indented at column 4 (1 standard indent). I can still try to make methods chains less bad.

**Comment by Christopher-Chianelli** [comment]
@yurloc @ge0ffrey should I address the formatting issue in another PR? It added/deleted 32,000 lines... I cannot fix the eclipse formatter file, so I am just running IDEA formatter in command line.

**Comment by hasys** [comment]
Hi @yurloc @Christopher-Chianelli,

it looks like this repo isn't formatted according to latest changes. So if you want to split formatting from changes you have two options:

1. don't use formater for this repo
2. format entire repo as one PR and send you changes as separate PR

File: employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/common/TableRow.java [hunk 6]
```diff
@@ -18,28 +18,30 @@
...
+import org.jboss.errai.ui.client.local.api.IsElement;
 import org.jboss.errai.ui.shared.api.annotations.DataField;
 import org.jboss.errai.ui.shared.api.annotations.EventHandler;
 import org.jboss.errai.ui.shared.api.annotations.ForEvent;
 import org.optaweb.employeerostering.gwtui.client.popups.ErrorPopup;
 import org.optaweb.employeerostering.gwtui.client.util.CommonUtils;
 
-public abstract class TableRow<T> extends Composite implements TakesValue<T> {
+public abstract class TableRow<T> implements TakesValue<T>,
+                                             IsElement {
```
[the remark this answers is not in the document]
Comment (hasys) [reply employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/common/TableRow.java]: @yurloc I guess this one is Eclipse/IDEA incompatibilities

[the remark this answers is not in the document]
Comment (Christopher-Chianelli) [reply employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/common/TableRow.java]: @yurloc IDEA, I ran the IDEA formatter from commandline on all source files in the last commit

**Comment by Christopher-Chianelli** [comment]
@yurloc In regards to the method chaining, I might be able to fix that in the Eclipse formatter, or at least change it to something more like IDEA's (since it is just "increase the indentation level by 1").

**Comment by Christopher-Chianelli** [comment]
@ge0ffrey Fixed i18n and bulk delete issue
